File: linear-and-multivariable-algebra/programs/projection.cpp
#include "linalg.hpp"

int main() {
    linalg::Vector a = {1, 2, 3}; // vector to project
    linalg::Vector b = {1, 0, 0}; // line to project onto
    
    double dot_ab = linalg::dot(a, b);
    double dot_bb = linalg::dot(b, b);
    
    double scalar = dot_ab / dot_bb;
    linalg::Vector proj(b.size());
    for(size_t i=0; i<b.size(); ++i) proj[i] = scalar * b[i];
    
    linalg::print(proj, "Projection of a onto b");
    return 0;
}

File: linear-and-multivariable-algebra/programs/projection.py
import numpy as np

def projection(u, v):
    # Proj of u onto v
    factor = np.dot(u, v) / np.dot(v, v)
    return factor * v

u = np.array([2, 1])
v = np.array([-3, 4])
print(projection(u, v))
